File: 2024_12/leetcode2024-12-12.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <climits>
using namespace std;

class Solution {
public:
    long long maxSpending(vector<vector<int>>& values) {
        vector<int> totVal;
        int m = values.size(), n = values[0].size();
        for (auto i : values){
            totVal.insert(totVal.end(), i.begin(), i.end());
        }
        sort(totVal.begin(), totVal.end());
        long long res = 0;
        for (int i=0; i<m*n; i++){
            res += 1ll*(i+1)*totVal[i];
        }
        return res;
        // long long res = 0, day = 1;
        // int m = values.size(), n = values[0].size();
        // vector<int> indx(m, -1);
        // for (int i=0; i<n*m; i++){
        //     int tmpIndx = 0, v = INT_MAX;
        //     for (int j=0; j<m; j++){
        //         if ((indx[j] != (-n-1)) && (values[j][indx[j]+n] < v)){
        //             v = values[j][indx[j]+n];
        //             tmpIndx = j;
        //         }
        //     }
        //     res += values[tmpIndx][indx[tmpIndx]+n]*day;
        //     day++;
        //     indx[tmpIndx]--;
        // }
        // return res;
    }
};

int main(){
    Solution s;
    vector<vector<int>> values = {{8,5,2},{6,4,1},{9,7,3}};
    cout << s.maxSpending(values) << endl;
}

File: 2024_12/leetcode2024-12-12.py
from typing import List

class Solution:
    def maxSpending(self, values: List[List[int]]) -> int:
        totVal = []
        for i in values:
            totVal.extend(i)
        totVal.sort()
        res = 0
        for i,v in enumerate(totVal):
            res += (i+1)*v
        return res
        # res,day = 0,1
        # n,m = len(values[0]),len(values)
        # indx = [-1]*m
        # for i in range(n*m):
        #     tmpIndx,v = 0,inf
        #     for j in range(0,m):
        #         if indx[j] != -n-1 and values[j][indx[j]] < v:
        #             v = values[j][indx[j]]
        #             tmpIndx = j
        #     res += values[tmpIndx][indx[tmpIndx]]*day
        #     day += 1
        #     indx[tmpIndx] -= 1
        # return res

if __name__ == "__main__":
    s = Solution()
    values = [[8,5,2],[6,4,1],[9,7,3]];
    print(s.maxSpending(values))
